pkg/filter/wasm/wasm.go

```diff
@@ -0,0 +1,317 @@
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package wasm
+
+import (
+	"bytes"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/megaease/easegress/pkg/cluster"
+	"github.com/megaease/easegress/pkg/context"
+	"github.com/megaease/easegress/pkg/logger"
+	"github.com/megaease/easegress/pkg/object/httppipeline"
+)
+
+const (
+	// Kind is the kind of WasmFilter.
+	Kind        = "WasmFilter"
+	maxWasmCode = 9
+)
+
+var (
+	resultOutOfVM   = "outOfVM"
+	resultWasmError = "wasmError"
+	results         = []string{resultOutOfVM, resultWasmError}
+)
+
+func wasmCodeToResult(code int32) string {
+	if code == 0 {
+		return ""
+	}
+	return fmt.Sprintf("wasmResult%d", code)
+}
+
+func init() {
+	for i := int32(1); i <= maxWasmCode; i++ {
+		results = append(results, wasmCodeToResult(i))
+	}
+	httppipeline.Register(&WasmFilter{})
+}
+
+type (
+	Spec struct {
+		MaxConcurrency int32  `yaml:"maxConcurrency" jsonschema:"required,minimum=1"`
+		Code           string `yaml:"code" jsonschema:"required"`
+		Timeout        string `yaml:"timeout" jsonschema:"required,format=duration"`
+		timeout        time.Duration
+	}
+
+	WasmFilter struct {
+		pipeSpec *httppipeline.FilterSpec
+		spec     *Spec
+
+		code   []byte
+		vmPool atomic.Value
+		chStop chan struct{}
+	}
+)
+
+// Kind returns the kind of WasmFilter.
+func (f *WasmFilter) Kind() string {
+	return Kind
+}
+
+// DefaultSpec returns the default spec of WasmFilter.
+func (f *WasmFilter) DefaultSpec() interface{} {
+	return &Spec{
+		MaxConcurrency: 10,
+		Timeout:        "100ms",
+	}
+}
+
+// Description returns the description of WasmFilter
+func (f *WasmFilter) Description() string {
+	return "WasmFilter implements a filter which runs web assembly"
+}
+
+// Results returns the results of WasmFilter.
+func (f *WasmFilter) Results() []string {
+	return results
+}
+
+func readWasmCodeFromFile(path string) ([]byte, error) {
+	f, e := os.Open(path)
+	if e != nil {
+		return nil, e
+	}
+	defer f.Close()
+	return io.ReadAll(f)
+}
+
+func readWasmCodeFromURL(url string) ([]byte, error) {
+	resp, e := http.DefaultClient.Get(url)
+	if e != nil {
+		return nil, e
+	}
+	defer resp.Body.Close()
+	return io.ReadAll(resp.Body)
+}
+
+func isURL(str string) bool {
+	// we only check the first a few bytes as str could be
+	// the LONG base64 encoded wasm code
+	for _, p := range []string{"http://", "https://"} {
```

Comment:
if `str` is the base64 encoded wasm code, it can be very long, but we only need to check the first a few bytes.

Reply:
> it will try to split the URL with `#`

this is the case, `#` is not a valid character in base64 encoding, so the split has to go through the whole string before it finds out that the string does need a split.

Reply:
Yes, if we care about the parsing cost especially here, your solution's performance is better than the standard lib.